FlexMeter.c

```diff
@@ -0,0 +1,216 @@
+   if (!strncmp(line, "name=", 5)) {
+      meter_list[meters_count].uiName = xStrdup(line + 5);
+   } else if (!strncmp(line, "command=", 7)) {
+      meter_list[meters_count].command = xStrdup(line + 8);
+      // DRY run command if it return err does not list the metter
+      char buff[256] = {0};
+      int ret = -1;
+      memset(buff, 0, 256);
+      FILE* fd = popen(meter_list[meters_count].command, "r");
+      if (fd) {
+         if (fgets(buff, 256, fd) == NULL) {
+            ret = pclose(fd);
+         } else {
+            ret = pclose(fd);
+         }
+      }
+
+      if (ret != 0) {
+         free(meter_list[meters_count].command);
+         meter_list[meters_count].command = NULL;
+      }
+   }
+   else if (!strncmp(line, "caption=", 7)) {
+      meter_list[meters_count].caption = xStrdup(line + 8);
+   } else if (!strncmp(line, "type=", 5)) {
+      meter_list[meters_count].type = xStrdup(line + 6);
+   } else {
+      return -1;
+   }
+
+   return 0;
+}
+
+static int load_config(char* file)
+{
+   int ret = -1;
+   char* buff;
+   FILE* fp = fopen( file, "r");
+
+   if (fp != NULL) {
+      while (1) {
+         buff = String_readLine(fp);
+         if (buff != NULL) {
+            ret = parse_input(buff);
+            if (ret) {
+               break;
+            }
+         }
+         else {
+            break;
+         }
+      }
+   }
+
+   fclose(fp);
+   return ret;
+}
+
+int check_for_meters(void)
+{
+   char path[400]; // full path
+   DIR* d;
+   struct dirent* dir;
+   struct passwd* pw = getpwuid(getuid());
+   const char* homedir = pw->pw_dir;
+
+   // path to home folder 1 for zero 1 for slash
+   char* home = (char* ) xCalloc(1, (strlen(homedir) + strlen(FLEX_CFG_FOLDER) + 2));
+
+   xSnprintf(home, (strlen(homedir) + strlen(FLEX_CFG_FOLDER) + 2), "%s/%s", homedir, FLEX_CFG_FOLDER);
+
+   struct stat fileStat;
+
+   if (stat(home, &fileStat) < 0) {
+      return -1;
+   }
+
+   if (S_ISDIR(fileStat.st_mode) &&
+         (fileStat.st_mode & S_IRUSR) &&
+         (fileStat.st_mode & S_IWUSR) &&
+         (fileStat.st_mode & S_IXUSR) ) {
+      d = opendir(home);
+      if (d) {
+         while ((dir = readdir(d)) != NULL) {
+            if ( dir->d_name[0] != '.') {
+               // We are ignoring all files starting with . like ".Template" and "." ".." directories
+               meter_list[meters_count].name = xStrdup(dir->d_name);
+               memset(path, 0, 80);
+               xSnprintf(path, 400, "%s/%s", home, dir->d_name);
+               if (access(path, R_OK | W_OK | X_OK) == 0) {
+                  if ( 0 != load_config(path) ) {
+                     break;
+                  }
+
+                  if ( meters_count < MAX_METERS_COUNT ) {
+                     meters_count++;
+                  } else {
+                     break; // go out we reach the limit
+                  }
+               }
+            }
+         }
+         closedir(d);
+      } else {
+         mkdir(home, 0700);
```

Comment:
Don't create folders when opening them fails.

Reply:
I am neutral on whether to create the folder anyway. If creating an empty folder would result in extra code then I would suggest to not bother with creating the folder.

Reply:
This is extra code, that is not needed when the folder exists. And if opening the folder fails this might be for other reasons than the folder being amiss. Thus the `mkdir` would need to be error-checked, which is code, we really don't need for the feature to work.

Reply:
this part is stripped now , if user need it, will create it